magicicadaprotocol/request.py

```diff
@@ -370,7 +370,7 @@ def cleanup(self):
         """remove the reference to self from the request handler"""
         self.finished = True
         self.started = False
-        del self.protocol.requests[self.id]
+        self.protocol.requests.pop(self.id, None)
```

Comment:
Here you're changing behaviour. In the new code you're ignoring the situation where self.id is not in the requests... this happens now and not before?

Reply:
Good catch. Actually this is a leftover of the test refactor for `test_client.py`, where I got rid of the arguably nasty `MethodMock` and added the `make_request` helper that returns a *real* request subclass instance.

With this change, the `request` that is being used in the tests, are real instances without mocked methods for `done` and `error`. So when the request is completed, either the real methods `done` or `error` are called which both call `cleanup`, which is the method you reference in your comment. The thing is that since no test (originally) called `start` on the request instances, the `id` for the instances is `None` and then we get a:

```
    self.protocol.requests.pop(self.id)
builtins.KeyError: None
```

Now that I explain this in writing, I think I'm going to try to call `start` in the test helper `make_request` and see if we get a cleaner code this way.

Reply:
+!